File: Applications/Examples/perl/market_price_authentication.pl
#|-----------------------------------------------------------------------------
#|            This source code is provided under the Apache 2.0 license      --
#|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
#|                See the project's LICENSE.md for details.                  --
#|           Copyright Thomson Reuters 2017. All rights reserved.            --
#|-----------------------------------------------------------------------------


#!/usr/bin/perl
# Simple example of outputting Market Price JSON data using Websockets with authentication

use feature qw(say);
use JSON;
use Mojo::UserAgent;
use Getopt::Long;
use Time::HiRes;
use Socket;
use Sys::Hostname;
use LWP::UserAgent;
use HTTP::Cookies;
use utf8;
use Encode;

# Global Default Variables
$hostname = "127.0.0.1";
$port = "15000";
$user = "root";
$app_id = "555";
$position = inet_ntoa(scalar(gethostbyname(hostname())) || 'localhost');
$auth_hostname = "127.0.0.1";
$auth_port = "8443";
$auth_token = "";
$help = "";

# Global Variables
$web_socket_open = 0;

use sigtrap 'handler' => \&myhand, 'INT';

sub myhand
{
    kill 6, $$; # ABRT = 6
}

# Get command line parameters
GetOptions ('hostname=s' => \$hostname, 'port=s' => \$port, 'user=s' => \$user, 'app_id=s' => \$app_id, 'position=s' => \$position, 'password=s' => \$password, 'auth_hostname=s' => \$auth_hostname, 'auth_port=s' => \$auth_port, 'help' => \$help);

if(not $help eq "") {
	print "Usage: market_price_authentication.pl [--hostname hostname] [--port port] [--app_id app_id] [--user user] [--password password] [--position position] [--auth_hostname auth_hostname] [--auth_port auth_port] [--help]\n";
	exit 0;
}

# Parse at high level and output JSON of message
sub process_message {
	my ($tx, $message_json) = @_;
	
	my $message_type = $message_json->{'Type'};
	
	if ($message_type eq "Refresh") {
		my $message_domain = $message_json->{'Domain'};
		if (not($message_domain eq "")) {
			if ($message_domain eq "Login") {
				send_market_price_request($tx);
			}
		}
	} elsif ($message_type eq "Ping") {

		my %pong_json_hash = (
			'Type' => 'Pong'
		);

		my $json = encode_json \%pong_json_hash;
		my $json_pretty = JSON->new->pretty->encode(\%pong_json_hash);
		$tx->send($json);
		print "SENT:\n";
		print "$json_pretty\n";
	}
}

# Create and send simple Market Price request
sub send_market_price_request {
	my ($tx) = @_;
	my %mp_req_json_hash = (
        'ID' => 2,
        'Key' => {
            'Name' => 'TRI.N',
        },
	);

	my $json = encode_json \%mp_req_json_hash;
	my $json_pretty = JSON->new->pretty->encode(\%mp_req_json_hash);
	$tx->send($json);
	print "SENT:\n";
	print "$json_pretty\n";
}

# Send login info for authentication token

print "Sending authentication request...\n";
use LWP::UserAgent;
my $ua = LWP::UserAgent->new;
$ua->agent("LWP::UserAgent (Perl)");
my $cookie_jar = HTTP::Cookies->new();
$ua->cookie_jar( $cookie_jar );
my $req = HTTP::Request->new(POST => "https://$auth_hostname:$auth_port/getToken");
$req->content_type('application/x-www-form-urlencoded');
$req->content("username=$user&password=$password");
$ua->ssl_opts( verify_hostname => 0 ,SSL_verify_mode => 0x00); 

# Pass request to the user agent and get a response back
my $res = $ua->request($req);
$cookie_jar->extract_cookies( $res );

# Check the outcome of the response
if ($res->is_success) {
	$res_json = decode_json $res->content;
	my $json_pretty = JSON->new->pretty->encode($res_json);
	print "RECEIVED:\n";
	print $json_pretty;
	
	if( $res_json->{'success'} ) {
	
		$cookie_jar->scan(sub  
        {  
          if ($_[1] eq "AuthToken") 
          { 
            $auth_token = $_[2];
          };
        });
		
		print("Authentication Succeeded. Received AuthToken: $auth_token\n");
		
		# Start websocket handshake
		my $ua = Mojo::UserAgent->new;
		
		$ua->on(start => sub {
		  my ($ua, $tx) = @_;
		  $tx->req->headers->header('Cookie' => "AuthToken=$auth_token;AuthPosition=$position;applicationId=$app_id;");
		});
		
		my $ws_address = "ws://$hostname:$port/WebSocket";
		print "Connecting to WebSocket $ws_address ...\n";
		$ua->inactivity_timeout(0);
		$ua->websocket($ws_address => ['tr_json2'] => sub {
			my ($ua, $tx) = @_;
			say 'WebSocket handshake failed!' and return unless $tx->is_websocket;
			say 'Subprotocol negotiation failed!' and return unless $tx->protocol;
			# Called when websocket is closed
			$tx->on(finish => sub {
				my ($tx, $code, $reason) = @_;
				say "WebSocket closed with status $code.";
			});
			# Called when message received, parse message into JSON for processing
			$tx->on(message => sub {
				my ($tx, $message) = @_;
				print "RECEIVED:\n";
				my @json_array = @{decode_json encode_utf8($message)};

				my $json_pretty = JSON->new->pretty->encode(\@json_array);
				print(encode_utf8($json_pretty));

				foreach my $single_message (@json_array) {
				process_message($tx, $single_message);
				}
			});
			
			print "WebSocket successfully connected!\n";
			$web_socket_open = 1;
		});

		# Event loop
		Mojo::IOLoop->start unless Mojo::IOLoop->is_running;
	}
	else {
		print("Authentication failed");
	}
}
else {
	print("Token request failed");
}

File: Applications/Examples/ruby/market_price_authentication.rb
#|-----------------------------------------------------------------------------
#|            This source code is provided under the Apache 2.0 license      --
#|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
#|                See the project's LICENSE.md for details.                  --
#|           Copyright (C) 2019-2020 Refinitiv. All rights reserved.         --
#|-----------------------------------------------------------------------------


#!/usr/bin/ruby
# * Simple example of outputting Market Price JSON data using Websockets with authentication

require 'rubygems'
require 'websocket-client-simple'
require 'json'
require 'optparse'
require 'socket'
require 'http'

# Global Default Variables
$hostname = '127.0.0.1'
$port = '15000'
$user = 'root'
$app_id = '555'
$position = Socket.ip_address_list[0].ip_address
$auth_hostname = '127.0.0.1'
$auth_port = '8443'
$user = ''
$password = ''
$auth_token = ''

# Get command line parameters
opt_parser = OptionParser.new do |opt|

  opt.on('--hostname HOST','HOST') do |hostname|
    $hostname = hostname
  end

  opt.on('--port port','port') do |port|
    $port = port
  end
  
  opt.on('--app_id APP_ID','APP_ID') do |app_id|
    $app_id = app_id
  end
  
  opt.on('--user USER','USER') do |user|
    $user = user
  end
  
  opt.on('--password PASSWORD','PASSWORD') do |password|
    $password = password
  end
  
  opt.on('--position POSITION','POSITION') do |position|
    $position = position
  end
  
  opt.on('--auth_hostname AUTH_HOSTNAME','AUTH_HOSTNAME') do |auth_hostname|
    $auth_hostname = auth_hostname
  end
  
  opt.on('--auth_port AUTH_PORT','AUTH_PORT') do |auth_port|
    $auth_port = auth_port
  end
  
  opt.on('--help','HELP') do |help|
	puts 'Usage: market_price_authentication.rb [--hostname hostname] [--port port] [--app_id app_id] [--user user] [--password password] [--position position] [--auth_hostname auth_hostname] [--auth_port auth_port] [--help]'
	exit 0
  end
end

opt_parser.parse!

# Create and send simple Market Price request
def send_market_price_request(ws)
  mp_req_json_hash = {
    'ID' => 2,
    'Key' => {
      'Name' => 'TRI.N'
    }
  }
  ws.send mp_req_json_hash.to_json.to_s
  puts 'SENT:'
  puts JSON.pretty_generate(mp_req_json_hash) 
end

# Parse at high level and output JSON of message
def process_message(ws, message_json)
  message_type = message_json['Type']
  
  if message_type == 'Refresh' then
    message_domain = message_json['Domain']
	if message_domain != nil then
	  if message_domain == 'Login' then
	    send_market_price_request(ws)
	  end
	end
  elsif message_type == 'Ping' then
    pong_json_hash = {
	    'Type' => 'Pong',
    }
    ws.send pong_json_hash.to_json.to_s
    puts 'SENT:'
    puts JSON.pretty_generate(pong_json_hash)  
  end
end

# Send login info for authentication token
puts 'Sending authentication request...'
headers = {'Content-Type' => 'application/x-www-form-urlencoded'}
url = "https://#{$auth_hostname}:#{$auth_port}/getToken"
data = "username=#{$user}&password=#{$password}"
ctx = OpenSSL::SSL::SSLContext.new
ctx.verify_mode = OpenSSL::SSL::VERIFY_NONE
auth_resp = HTTP.headers(headers).post(url, body: data, ssl_context: ctx)
auth_json = JSON.parse(auth_resp.to_s)

puts 'RECEIVED:'
puts JSON.pretty_generate(auth_json)
 
if auth_json['success'] == true
  # Start websocket handshake
  ws_address = "ws://#{$hostname}:#{$port}/WebSocket"
  $auth_token = HTTP::Cookie.cookie_value_to_hash(HTTP::Cookie.cookie_value(auth_resp.cookies.cookies))["AuthToken"]
  puts "Authentication Succeeded. Received AuthToken: #{$auth_token}"
  puts "Connecting to WebSocket #{ws_address} ..."
  ws = WebSocket::Client::Simple.connect(ws_address,{headers: {'Sec-WebSocket-Protocol' => 'tr_json2', 'User-Agent' => 'Ruby', 'Cookie' => "AuthToken=#{$auth_token};AuthPosition=#{$position};applicationId=#{$app_id};"}})
else
  puts "Authentication failed"
  exit -1
end

# Called when message received, parse message into JSON for processing
ws.on :message do |msg|
  msg = msg.to_s

  puts 'RECEIVED:'

  json_array = JSON.parse(msg)

  puts JSON.pretty_generate(json_array)

  for single_msg in json_array
    process_message(ws, single_msg)
  end
  
end

# Called when handshake is complete and websocket is open, send login
ws.on :open do
  puts 'WebSocket successfully connected!'
end

# Called when websocket is closed
ws.on :close do |e|
  puts 'CLOSED'
  p e
  exit 1
end

# Called when websocket error has occurred 
ws.on :error do |e|
  puts 'ERROR'
  p e
end

sleep
